### DISCUSSION TITLE
question about implements LRU policy

### DISCUSSION THREAD

**Discussion by xiaoxiangshusheng** [InternLM/lmdeploy#2316] [Q&A] [answered]
In turbomind.md, turboMind is a memory-pool-liked object that also implements LRU policy so that it can be viewed as a form of __cache of KV caches__. However, the code in lmdeploy seems to use LFU.
![1723705848882](https://github.com/user-attachments/assets/59384cf0-4ce4-4ec9-a4fd-8241c1f83d9a)

Althrough return blocks_[i].timestamp < blocks_[j].timestamp in code,  timestamp is counted by the cumulative form . In other words, timestamp is not a representation of the current time. In code, the timestamp will be added 1 if the blocks is touched. So could you 
 help me to answer my confusion ablout LRU policy to delete evict blocks?

**Comment by lzhangzz** [comment, marked as answer]
Even though timestamp is cumulative, `blocks_[i].timestamp < blocks_[j].timestamp` still implies `i` is used less recently than `j`.

Notice that the blocks of a sequence is touched in reversed order. This enforce the blocks in the back of a sequence is evicted before blocks in the front of the sequence.
